cpp/src/arrow/compute/exec/aggregate.cc

```diff
@@ -38,33 +38,38 @@ namespace internal {
 
 Result<std::vector<const HashAggregateKernel*>> GetKernels(
     ExecContext* ctx, const std::vector<Aggregate>& aggregates,
-    const std::vector<TypeHolder>& in_types) {
+    const std::vector<std::vector<TypeHolder>>& in_types) {
   if (aggregates.size() != in_types.size()) {
     return Status::Invalid(aggregates.size(), " aggregate functions were specified but ",
                            in_types.size(), " arguments were provided.");
   }
 
   std::vector<const HashAggregateKernel*> kernels(in_types.size());
 
+  std::vector<TypeHolder> aggregate_in_types;
```

Comment:
maybe reserve size to the vector?

Reply:
Given that this is the loop body:

```cpp
      // {in_types[i]..., uint32()}
      aggregate_in_types = in_types[i];
      aggregate_in_types.emplace_back(uint32());
```

do you prefer a `.reserve(2)` for the common case of `in_types[i].size() == 1` or should I reserve within the loop body like this?

```diff
       // {in_types[i]..., uint32()}
+      aggregate_in_types.reserve(in_types[i].size() + 1);
       aggregate_in_types = in_types[i];
       aggregate_in_types.emplace_back(uint32());
```

Reply:
Sorry if I misread this, if it’s hard to guess the exact value, let’s keep it as it is. 

cc @westonpace thoughts?

Reply:
I'm adding a commit fixing both cases using the second approach.